### DISCUSSION TITLE
Clarify CalendarEntity timezone handling obligations for reads

### DISCUSSION THREAD

**Discussion by allenporter** [home-assistant/architecture#853] [Entity Models] [answered]
I would like to clarify `CalendarEvent` timezone handling requirements during reads to be more defensive against bugs, and to fix some inconsistencies between the developer guidance and implementations today.

## Background

The [CalendarEntity](https://developers.home-assistant.io/docs/core/entity/calendar) has two responsibilities it needs to fulfill: (1) Returning a `CalendarEvent` in a [time range](https://developers.home-assistant.io/docs/core/entity/calendar#get-events) and (2) Returning the [next or currently active](https://developers.home-assistant.io/docs/core/entity/calendar#properties) `CalendarEvent`. We do not currently have timezone handling guidance, which is an important part of implementing these APIs correctly.

## Overview

Independent of Home Assistant, the types of calendar event start date/date-times ([rfc5545](https://www.rfc-editor.org/rfc/rfc5545#section-3.3.5)) to consider are: 
- Events in UTC or with a specific Timezone (these don't need discussion here)
- "All day" events and "floating" date-time events. Note: An all day event is basically like a floating date-time event at midnight.

When mapping these calendar native constructs to home assistant APIs, a calendar integration is responsible for considering events relative to timezone to order their events properly. (e.g. mixing all day events and timed events)

## Proposed Guidance

- When returning active or upcoming events, integrations should use `homeassistant.util.dt.now()` and map all day/floating events based on the `tzinfo` of the returned `datetime`
- When scanning for a range of events in `async_get_events`, integrations should use the `start` value `tzinfo` to determine the timezone for all day/floating events. The `start` and `end` will have the same `tzinfo`.
- Returned `CalendarEvent` with a `datetime` must have a timezone, relaxed from requiring UTC. Unchanged: No floating events are allowed in this context, and all day events are still allowed as `date`.

## Proposed Work

- Add validation logic in the `CalendarEvent` dataclass to enforce time zones are present, and update all violating core integrations (e.g. caldav, local calendar)
- Update call sites of `async_get_events` to pass in local start/end dates
- Streamline existing integrations to use the start/end dates for timezones
- Developer blog post

## Alternatives

An alternative is to continue to use UTC across all of these APIs and:
- Only update docs that explain how the integration should know about `homeassistant.util.dt.DEFAULT_TIME_ZONE`, `dt.as_local`, and `dt.start_of_local_day`, etc.
- Enforce all integrations convert their `CalendarEvent` times to UTC and drop timezone information. Note: The 3 most popular calendar integrations currently don't return UTC times today as a datapoint

## Consequences

Many other APIs in Home Assistant require UTC only, since they are used for things like persisting timestamp states, and this will be different. I think a calendar is not a place where we want to set a precedent that we can just drop or ignore timezone information as it will lead to bugs.

My impression is doing comparisons in local time instead of UTC should be the same since everything is converted to UTC behind the scenes. (e.g. updating trigger call sites to use local datetime schedule should be a no-op)

Triggers for calendar events will be easier to debug by humans when everything is in local time.

**Comment by MartinHjelmare** [comment, marked as answer]
Sounds good! I'd like at least another member's opinion.

**Reply by balloob** [reply]
Yeah this is good.
